collectd/src/main/java/com/microsoft/applicationinsights/collectd/ApplicationInsightsWriter.java

```diff
@@ -0,0 +1,227 @@
+                if (values.size() != 1) {
+                    logger.logError(key + " configuration option needs exactly 1 argument [true/false].");
+                    return 2;
+                }
+
+                boolean enableSDKLogger = Boolean.parseBoolean(values.get(0).toString());
+                if (enableSDKLogger) {
+                    InternalLogger.INSTANCE.initialize("CONSOLE", new HashMap<String, String>());
+                }
+            } else {
+                logger.logWarning("Unknown configuration option '" + key + "'.");
+            }
+        }
+
+        logger.logInfo("Configuration loaded successfully.");
+
+        return 0;
+    }
+
+    /**
+     * Initializes the writer plugin.
+     * @return Zero on success, non-zero otherwise.
+     */
+    public int init() {
+        logger.logInfo("Initializing...");
+
+        try {
+            this.telemetryClient = new TelemetryClient(this.telemetryConfiguration);
+
+            logger.logInfo("Initialization completed.");
+        } catch (Throwable e ) {
+            logger.logError("Initialization failed, plugin will be disabled:\n" + e.toString());
+
+            return 1;
+        }
+
+        return 0;
+    }
+
+    /**
+     * Shutting down the writer plugin.
+     * @return Zero on success, non-zero otherwise.
+     */
+    public int shutdown() {
+        return 0;
+    }
+
+    /**
+     * Writes the given values to Application Insights.
+     * @param valueList The values to be sent.
+     * @return Zero on success, non-zero otherwise.
+     */
+    public int write(ValueList valueList) {
+        // We're avoiding writing traces in this method to not spam the logs, as this method is being called very frequently.
+        logger.logDebug("Writing values (" + valueList.getValues().size() + "):\n" + valueList.toString());
+
+        for (int i = 0; i < valueList.getValues().size(); i++) {
+            MetricTelemetry metricTelemetry = createMetricTelemetry(valueList, i);
+            this.telemetryClient.trackMetric(metricTelemetry);
+        }
+
+        return 0;
+    }
+
+    // endregion Public
+
+    // region Private
+
+    private static MetricTelemetry createMetricTelemetry(ValueList valueList, int index) {
+        DataSource dataSource = getDataSource(valueList.getDataSet(), index);
+
+        MetricTelemetry telemetry = new MetricTelemetry();
+
+        if (dataSource != null) {
+            telemetry.setMin(dataSource.getMin());
+            telemetry.setMax(dataSource.getMax());
+        }
+
+        String telemetryName = generateMetricName(valueList, dataSource);
+        telemetry.setName(telemetryName);
+
+        Number value = valueList.getValues().get(index);
+        if (value != null) {
+            telemetry.setValue(value.doubleValue());
+        }
+
+        return telemetry;
+    }
+
+    protected static String generateMetricName(ValueList valueList, DataSource dataSource) {
```

Comment:
Is it possible to somehow cache these values?
I'm asking since in each write we're constructing probably the same metric names over and over, so this sounds as an inefficient process.